### ISSUE TITLE
Selector and Model fail to return parameters silently

### ISSUE THREAD

**Issue by bartvm** [bartvm/blocks#480]
If parameters have the same path/name (i.e. if a sequence contains multiple `Linear` bricks without a specific name) the `Selector.get_params` (and hence `Model.get_parameters`) will simply override the first one with the second one in the parameter dictionary it returns, and hence fail to return all the parameters.

**Comment by rizar** [comment]
My opinion is that we should enforce that children of a brick have different names. You `Children` class is a good place to check that.

**Comment by bartvm** [comment]
Sounds good. The same problem occurs for parameters outside of the brick hierarchy though. The `Model` class simply stores them with the `param.name` key, which could easily be the same for multiple parameters, and there too it silently doesn't return some parameters.

**Comment by rizar** [comment]
Right, we should as well check uniqueness in both Model and Selector.

**Comment by dwf** [comment]
I saw @mducoffe run into an error that sounds like @rizar's suggestion. Has this been addressed/should this ticket be closed?

**Comment by mducoffe** [comment]
@ rizar Is there any way to see which variable is concerned by the problem. I have put names to everything during the construction but I keep having the problem. What is weird is that my cost is composed of two cost : cost_1 + cost_2. When I am considering only cost_1 I have the error ( so I can't save my model) but when I am running the sum of cost_1 and cost_2 it is working so I am a bit perplexed.

**Comment by rizar** [comment]
If you are talking about "top bricks have the same name", then it's not about variable and about brick names. Can you please show the code? You can use `gist.github,com` for that.

**Comment by rizar** [comment]
@dwf it's only `Model` that does this check so far,  we need it to be also done in `Children` and `Selector`.

**Comment by bartvm** [comment]
This is just a note that @vdumoulin's solution in #597 only applies to parameters that belong to bricks. In my case I replace the parameters in the model with Gaussians using `ComputationGraph.replace`, which introduces a whole bunch of new parameters (the variances) which don't belong to any brick. These still get silently overwritten in [this line](https://github.com/bartvm/blocks/blob/master/blocks/model.py#L171) and never get saved.
